search/binarySearchTree.zig

```diff
@@ -1,82 +1,241 @@
+                return root;
+            }
+
+            if (root.?.info < key) {
+                root.?.right = try self._remove(root.?.right, key);
+            } else if (root.?.info > key) {
+                root.?.left = try self._remove(root.?.left, key);
             } else {
-                y.?.right = z;
+                if (root.?.left == null and root.?.right == null) {
+                    self.allocator.destroy(root.?);
+                    return null;
+                } else if (root.?.left == null) {
+                    const temp = root.?.right;
+                    self.allocator.destroy(root.?);
+                    return temp;
+                } else if (root.?.right == null) {
+                    const temp = root.?.left;
+                    self.allocator.destroy(root.?);
+                    return temp;
+                } else {
+                    var curr: ?*node = root.?.right;
+                    while (curr.?.left != null) : (curr = curr.?.left) {}
+                    root.?.info = curr.?.info;
+                    root.?.right = try self._remove(root.?.right, curr.?.info);
+                }
+            }
+
+            return root;
+        }
+
+        fn _search(root: ?*node, key: T) bool {
+            var head: ?*node = root;
+            while (head) |curr| {
+                if (curr.info < key) {
+                    head = curr.right;
+                } else if (curr.info > key) {
+                    head = curr.left;
+                } else {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        fn _inorder(self: *Self, root: ?*node, path: *ArrayList(T)) !void {
+            if (root != null) {
+                try self._inorder(root.?.left, path);
+                try path.append(root.?.info);
+                try self._inorder(root.?.right, path);
+            }
+        }
+
+        fn _preorder(self: *Self, root: ?*node, path: *ArrayList(T)) !void {
+            if (root != null) {
+                try path.append(root.?.info);
+                try self._preorder(root.?.left, path);
+                try self._preorder(root.?.right, path);
+            }
+        }
+
+        fn _postorder(self: *Self, root: ?*node, path: *ArrayList(T)) !void {
+            if (root != null) {
+                try self._postorder(root.?.left, path);
+                try self._postorder(root.?.right, path);
+                try path.append(root.?.info);
+            }
+        }
+
+        fn _destroy(self: *Self, root: ?*node) void {
+            if (root != null) {
+                self._destroy(root.?.left);
+                self._destroy(root.?.right);
+                self.allocator.destroy(root.?);
             }
         }
     };
 }
 
-test "search empty tree" {
-    const tree = Tree(i32){};
-    const result = Tree(i32).search(tree.root, 3);
-    try expect(result == null);
-}
+test "Testing Binary Search Tree" {
```

Comment:
Please! Break down or add more tests to evaluate the run-time of each operation. Like the previous example, following the zig-std tests format.